oasis/models/rules.py

```diff
@@ -2,15 +2,167 @@
         self.api = PrometheusAPI(job.data_source)
+        self.timer = Timer(timeparse(self.job.timeout), self.timeout_action)
+        self.config_file = "%s/rules.yml" % self.model_path
+        self.cfg = RulesConfig(self.config_file, job.config)
+
+    def run(self):
+        self.timer.start()
+        for key in self.job.metrics:
+            if key in Metrics:
+                val = Metrics[key]
+                if key not in self.cfg.metrics:
+                    continue
+
+                t = Thread(target=self.run_action,
+                           args=(key, val, self.cfg.metrics[key],
+                                 self.cfg.rules[key], ))
+                t.start()
+                self.threads[key] = t
+
+    def run_action(self, metric, val, config, rules):
+        logger.info("[job-id:{id}][metric:{metric}] start ot run"
+                    .format(id=sub_id(self.job.id), metric=metric))
+        self.compute(metric, val, config, rules)
+
+    def compute(self, metric, query_expr, config, rules):
+        while not self._exit:
+            now = datetime.datetime.now()
+            query = PrometheusQuery(query_expr,
+                                    time.mktime((now - datetime.timedelta(
+                                        seconds=timeparse(self.cfg.model["data_range"])))
+                                                .timetuple()),
+                                    time.mktime(now.timetuple()), "15s")
+            data_set = self.api.query(query)
+            if len(data_set) > 0:
+                values = []
+                for data in data_set.values():
+                    values.append(float(data))
+
+                features_value = {}
+                for key in config["features"]:
+                    if key in Features:
+                        features_value[key] = Features[key](values)
+
+                logger.info("[job-id:{id}][metric:{metric}] extraction features {value}, "
+                            "start to compare with rule"
+                            .format(id=sub_id(self.job.id),
+                                    metric=metric, value=features_value))
+
+                not_match_count = 0
+                for rule in rules:
+                    if rule["feature"] not in features_value:
+                        continue
+
+                    if check_is_triggered(features_value[rule["feature"]],
+                                          rule["operator"], rule["value"]):
+                        logger.info("[job-id:{id}][metric:{metric}] not match rule: {rule}"
+                                    .format(id=sub_id(self.job.id), metric=metric, rule=rule))
+                        self.callback("[job] {job}, metric: {metric} not match rule: {rule}"
+                                      .format(job=dict(self.job), metric=metric,
+                                              rule=rule), self.job.slack_channel)
+                        not_match_count = not_match_count + 1
```

Comment:
Looks like if `not_match_count > 0`, we can break early here. So, something like `match_flag` would be more appropriate?